Libraries/LibWeb/DOM/Element.cpp

```diff
@@ -3333,61 +3336,63 @@ Element const* Element::list_owner() const
 }
 
 // https://html.spec.whatwg.org/multipage/grouping-content.html#ordinal-value
-size_t Element::ordinal_value() const
+WebIDL::Long Element::ordinal_value()
 {
-    // NOTE: The spec provides an algorithm to determine the ordinal value of each element owned by a given list owner.
-    //       However, we are only interested in the ordinal value of this element.
+    if (m_ordinal_value.has_value())
+        return m_ordinal_value.value();
 
-    // FIXME: 1. Let i be 1.
-
-    // 2. If owner is an ol element, let numbering be owner's starting value. Otherwise, let numbering be 1.
     auto const* owner = list_owner();
     if (!owner) {
         return 1;
     }
 
-    auto numbering = 1;
+    // 1. Let i be 1. [Not necessary]
+    // 2. If owner is an ol element, let numbering be owner's starting value. Otherwise, let numbering be 1.
+    WebIDL::Long numbering = 1;
     auto reversed = false;
     if (is<HTML::HTMLOListElement>(owner)) {
         auto const* ol_element = static_cast<const HTML::HTMLOListElement*>(owner);
         numbering = ol_element->starting_value();
         reversed = ol_element->has_attribute(HTML::AttributeNames::reversed);
-    }
 
-    // FIXME: 3. Loop : If i is greater than the number of list items that owner owns, then return; all of owner's owned list items have been assigned ordinal values.
-    // FIXME: 4. Let item be the ith of owner's owned list items, in tree order.
+        if (reversed)
+            VERIFY(AK::NumericLimits<WebIDL::Long>::min() + ol_element->number_of_owned_list_items() <= numbering);
+        else
+            VERIFY(AK::NumericLimits<WebIDL::Long>::max() - ol_element->number_of_owned_list_items() >= numbering);
+    }
 
+    // 3. Loop : If i is greater than the number of list items that owner owns, then return; all of owner's owned list items have been assigned ordinal values.
+    // NOTE: We use `owner->for_each_child_of_type` to iterate through the owner's list of owned elements.
+    //       As a result, we don't need `i` as counter (spec) in the list of children, with no material consequences.
     owner->for_each_child_of_type<DOM::Element>([&](auto& item) {
```

Comment:
I just noticed this mistake in the original implementation.

```suggestion
    owner->for_each_in_subtree_of_type<DOM::Element>([&](auto& item) {
```

Reply:
This clarifies  it _a lot_! I didn't question the original implementation; just looked at performance consequences.
Your comment before now also makes total sense.
Planning to add a few unit tests for this too.